model.go

```diff
@@ -436,3 +382,143 @@ func (m *Machine) changedAnnotations(annotations map[string]string) map[string]s
 	}
 	return changes
 }
+
+type inferenceEngine struct {
+	model  *Model
+	bundle *charm.BundleData
+
+	appUnits      map[string][]Unit
+	appPlacements map[string][]string
+
+	initialMachines set.Strings
+	logger          Logger
+}
+
+func newInference(m *Model, data *charm.BundleData) *inferenceEngine {
+	log := m.logger
+	if log == nil {
+		log = &noOpLogger{}
+	}
+	appUnits := make(map[string][]Unit)
+	// The initialMachines starts by including all the targets defined
+	// by the user for the machine map.
+	initialMachines := set.NewStrings()
+	for _, target := range m.MachineMap {
+		initialMachines.Add(target)
+	}
+	for appName, app := range m.Applications {
+		var units []Unit
+		// If the unit is on a machine we have already been told the mapping for,
+		// skip it in the inference.
+		for _, unit := range app.Units {
+			machine := topLevelMachine(unit.Machine)
+			if !initialMachines.Contains(machine) {
+				units = append(units, unit)
+			}
+		}
+		appUnits[appName] = units
+	}
+	return &inferenceEngine{
+		model:           m,
+		bundle:          data,
+		appUnits:        appUnits,
+		appPlacements:   make(map[string][]string),
+		initialMachines: initialMachines,
+		logger:          log,
+	}
+}
+
+func (e *inferenceEngine) initialPlacements() {
+	for appName, app := range e.bundle.Applications {
+		var unused []string
+		for _, to := range app.To {
+			unused = append(unused, to)
+			placement, _ := charm.ParsePlacement(to)
+			if placement == nil || placement.Machine == "" {
```

Comment:
Can we copy the comment from down below up here too. I was wondering why the error was ignored until I read that  comment.